### ISSUE TITLE
Segmentation fault

### ISSUE THREAD

**Issue by DaAngus** [CleverRaven/Cataclysm-DDA#29836]
**Describe the bug**  
`The program has crashed.
See the log file for a stack trace.
CRASH LOG FILE: ./config/crash.log
VERSION: 0.D-2101-g5f77bb325a
TYPE: St9bad_alloc
MESSAGE: std::bad_alloc
STACK TRACE:

	./cataclysm-tiles(_Z21debug_write_backtraceRSo+0x43) [0x5625cfc10bd1]
	./cataclysm-tiles(+0x403dd1) [0x5625cfc01dd1]
	./cataclysm-tiles(+0x403f7f) [0x5625cfc01f7f]
	/usr/lib/libstdc++.so.6(+0x8fdfa) [0x7f4faca07dfa]
	/usr/lib/libstdc++.so.6(+0x8fe57) [0x7f4faca07e57]
	/usr/lib/libstdc++.so.6(+0x900ac) [0x7f4faca080ac]
	/usr/lib/libstdc++.so.6(+0xb9748) [0x7f4faca31748]
	./cataclysm-tiles(+0x4441af) [0x5625cfc421af]
	./cataclysm-tiles(_ZNSt6vectorISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiESaIS7_EEC1ERKS9_+0x49) [0x5625cfc4e695]
	./cataclysm-tiles(_ZNK6effect13get_miss_msgsB5cxx11Ev+0x13) [0x5625cfc495b3]
	./cataclysm-tiles(_ZN6player18process_one_effectER6effectb+0x85) [0x5625d00718e9]
	./cataclysm-tiles(_ZN6player15process_effectsEv+0x3a2) [0x5625d008f124]
	./cataclysm-tiles(_ZN8Creature12process_turnEv+0x26) [0x5625cfc024c0]
	./cataclysm-tiles(_ZN6player12process_turnEv+0x2c) [0x5625d008d8b8]
	./cataclysm-tiles(_ZN4game7do_turnEv+0x731) [0x5625cfcb2f6d]
	./cataclysm-tiles(main+0x1230) [0x5625cfac11c4]
	/usr/lib/libc.so.6(__libc_start_main+0xf3) [0x7f4fac604223]
	./cataclysm-tiles(_start+0x2e) [0x5625cfaf36de]

	Attempting to repeat stack trace using debug symbols...
	debug_write_backtrace(std::ostream&)
	??:?
	log_crash
	crash.cpp:?
	crash_terminate_handler()
	crash.cpp:?
	??
	??:0
	??
	??:0
	??
	??:0
	??
	??:0
	__gnu_cxx::new_allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> >::allocate(unsigned long, void const*) [clone .isra.345]
	effect.cpp:?
	std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > >::vector(std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&)
	??:?
	effect::get_miss_msgs[abi:cxx11]() const
	??:?
	player::process_one_effect(effect&, bool)
	??:?
	player::process_effects()
	??:?
	Creature::process_turn()
	??:?
	player::process_turn()
	??:?
	game::do_turn()
	??:?
	main
	??:?
	__libc_start_main
	??:?
	_start
	??:?
`

**To Reproduce**  
1. Sleep with or without the soporific bionic for any amount of time. 
(prior to the crashes installed, not sure if related)
2. Wait until you wake up
3. Like in real life waking up is the problem

**Expected behavior**  
You feel well rested.

**Screenshots**  

**Versions and configuration(please complete the following information):**
 - OS: Arch linux
 - Game Version: VERSION: 0.D-2101-g5f77bb325a
 - Graphics version: [Tiles]
 - Mods loaded: 

**Additional context**  

World: Wells Branch, Character: Lachlan
It is only this Character in this World, the save is fairly old

**Comment by neitsa** [comment]
TL;DR: it's a crash on a low memory condition.

Looking at the  stack trace, which can be summarized as follow:

```
crash.cpp:?
__gnu_cxx::new_allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> >::allocate(unsigned long, void const*) [clone .isra.345]
effect.cpp:?
std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > >::vector(std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int> > > const&)
effect::get_miss_msgs[abi:cxx11]() const
player::process_one_effect(effect&, bool)
```

So basically it crashes in `effect::get_miss_msgs` which is a very simple function:

https://github.com/CleverRaven/Cataclysm-DDA/blob/e169df3755f3dc775d4c6518c87bafeee9aa53ca/src/effect.cpp#L1141-L1144

The game tried to allocate a `std::vector<std::pair<std::string, int>>` vector but simply failed to do so.

Given that the crash signal gives us [`std::bad_alloc`](https://en.cppreference.com/w/cpp/memory/new/bad_alloc) (this means the game wasn't able to get memory from the system when it requested for some) it really seems that you were running out of memory at that point. I can't see any other explanation.

Some suggestions: 

* If you are playing on 64-bit system with a 32-bit version of the game, please use a [64-bit one](http://dev.narc.ro/cataclysm/jenkins-latest/Linux_x64/).
* The game is memory hungry when using tilesets (memory usage goes often to approx. 2GB), you might consider using the curses version.
* Try to not run any other memory intensive process(es) while running the game.

@ifreund May I suggest to close the issue? there's not really much things we can do...

**Comment by kevingranade** [comment]
Two things, one specific and one general:
Can we catch the bad alloc and print a message?  This is one of the few failure conditions that users  can do something about if they're told.
We should do some memory profiling and see what comes up.  Map memory in particular has a disproportionately large memory footprint that can be reigned in.

**Comment by neitsa** [comment]
> Can we catch the bad alloc and print a message? This is one of the few failure conditions that users can do something about if they're told.

Yes Kevin, probably. 

My main concern is that the system is already in low memory condition and in this precise case we were lucky to get a stack trace. I guess (not sure though) that when `std::bad_alloc` is thrown and caught by the signal handler all the smart pointers (i.e `shared_ptr` and similar) are freed which gives some room to generate an error message.

Are you thinking about something specific, or just a simple message box like `"Your system is low on memory and Cataclysm: Dark Days Ahead can not continue to run. The game will now exit."` would be enough?

> We should do some memory profiling and see what comes up. Map memory in particular has a disproportionately large memory footprint that can be reigned in.

I have seen very recently - I'd say starting from 2 or 3 days ago - a visible increase in the memory footprint, (at least on the Tiles version). I don't have a very precise number but it was around 200 to 400 MB **on a debug version**.
That being said, I do agree with you that we need to profile the game and try to reduce the memory footprint. I can definitely help with this task.

**Comment by kevingranade** [comment]
"Your system is low on memory and Cataclysm: Dark Days Ahead can not
continue to run. The game will now exit." Is totally fine, ideally we can
have most resources allocated for this statically so it can present the
message without allocation.
